lib/src/deserializer.dart

```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// deserializer.dart
+// A deserialization block, deserializing narrow input data onto a wide channel.
+//
+// 2024 August 27
+// Author: desmond Kirkpatrick <desmond.a.kirkpatrick@intel.com>
+
+import 'package:meta/meta.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// Aggregates data from a serialized stream
+class Deserializer extends Module {
+  /// Clk input
+  @protected
+  Logic get clk => input('clk');
+
+  /// Reset input
+  @protected
+  Logic get reset => input('reset');
+
+  /// Run deserialization whenever [enable] is true
+  @protected
+  Logic? get enable => input('enable');
+
+  /// Serialized input, one data item per clock
+  Logic get serialized => input('serialized');
+
+  /// Aggregated data output
+  LogicArray get deserialized => output('deserialized') as LogicArray;
+
+  /// Valid out when data is reached
+  Logic get done => output('done');
+
+  /// Return the count as an output
+  @protected
+  Logic get count => output('count');
+
+  /// Build a Deserializer that takes serialized input [serialized]
+  /// and aggregates it into one wide output [deserialized], one element per
+  /// clock while [enable] (if connected) is high, emitting [done] when
+  /// completing the filling of wide output [deserialized].
+  Deserializer(Logic serialized, int length,
+      {required Logic clk,
+      required Logic reset,
+      Logic? enable,
+      super.name = 'Deserializer'}) {
+    clk = addInput('clk', clk);
+    reset = addInput('reset', reset);
+    if (enable != null) {
+      enable = addInput('enable', enable);
+    } else {
+      enable = Const(1);
+    }
+    serialized = addInput('serialized', serialized, width: serialized.width);
+    final cnt = Counter.simple(
+        clk: clk, reset: reset, enable: enable, maxValue: length - 1);
+    addOutput('count', width: cnt.width) <= cnt.count;
+    addOutput('done') <= cnt.overflowed;
```

Comment:
Is this right?  in the serializer you use `equalsMax`, and here you use `overflowed`.  I wonder if one of them is wrong?

It would also be good to be extremely clear in the doc comments for the `done` output when the user should expect the signal to go high (is it on the cycle *after* the last one was sent, or on the *same cycle* as the last one?)

Reply:
So I think I found an asymmetry.  Deserialization uses a flop by default to construct the aggregate (it could just mux latest input into the output port), but instead we are flopping.  Serialization has a conditional flop.

So the overflowed is really just a flop of equalsMax.    I think the fix is to create a deserialization that does NOT flop, and then flop conditionally the done and the aggregate array.

Reply:
I'm also thinking:  should we eliminate conditional flopping. I mean, if someone wants to flop the outputs of these components, then the 'done' signals should be the enables for those flops -- no reason to have the components do the flopping.

Reply:
Now that I look at it, it makes sense that deserialization must flop to accumulate the result and will necessarily take an extra cycle.